File: leetcode/contest-259/2014. Longest Subsequence Repeated k Times.cpp
class Solution {
    int k;
    unordered_map<char, int> cnt;
    string s, t, ans;

    void dfs() {
        unordered_map<char, int> cnt2(cnt);
        for (auto [ch, freq] : cnt2) {
            // 排列方式 t 给定了，判断是否出现了 k 次
            t.push_back(ch);
            if (t.size() > ans.size() || (t.size() == ans.size() && t > ans)) {
                int times = 0, pos = 0;
                for (char c : s)
                    if (c == t[pos]) {
                        pos++;
                        // 一次匹配完毕，times+=1,pos归零重新开始匹配
                        if (pos == t.size())
                            times++, pos = 0;
                    }

                if (times >= k)
                    ans = t;
            }
            // 出现次数减去 k
            cnt[ch] -= k;
            if (cnt[ch] < k)
                cnt.erase(ch);
            dfs();
            t.pop_back();
            cnt[ch] = freq;
        }
    }

   public:
    string longestSubsequenceRepeatedK(string s, int k) {
        this->k = k;
        int n = s.size();
        for (char c : s)
            cnt[c]++;
        string to_del;
        for (auto [ch, freq] : cnt)
            if (freq < k)
                to_del.push_back(ch);
        // 只留下出现次数 >= k 的那些字符，字符种数最多只有 7
        for (char ch : to_del)
            cnt.erase(ch);

        for (char c : s)
            if (cnt.count(c))
                this->s.push_back(c);

        dfs();

        return ans;
    }
};

File: leetcode/残酷刷题/2014. Longest Subsequence Repeated k Times.ts
function longestSubsequenceRepeatedK (s: string, k: number): string {
  const cnt = Array<number>(26).fill(0)
  for (let i = 0; i < s.length; i++) {
    cnt[s[i].charCodeAt(0) - 'a'.charCodeAt(0)] += 1
  }
  // c.length <= 7
  const c = new Map<number, number>()
  for (let i = 0; i < 26; i++) {
    if (cnt[i] >= k) c.set(i, cnt[i])
  }
  let allowed = ''; let curSeq = ''; let ans = ''
  for (const ch of s) if (c.has(ch.charCodeAt(0) - 'a'.charCodeAt(0))) allowed += ch
  dfs()
  return ans
  // O(t)
  function cntTimes (s: string, t: string) {
    let times = 0; let pos = 0
    for (const ch of t) {
      if (ch === s[pos]) {
        pos++
        if (pos === s.length) {
          times++; pos = 0
        }
      }
    }
    return times
  }
  // 枚举 c 中字符所有排列，O(7!)，每个结果需要和 allowed 比对，O(n=16000)
  function dfs () {
    const c1 = new Map(c)
    for (const [ch, cnt] of c1.entries()) {
      curSeq += String.fromCharCode(ch + 'a'.charCodeAt(0))
      if (curSeq.length > ans.length || (curSeq.length === ans.length && curSeq > ans)) {
        if (cntTimes(curSeq, allowed) >= k) ans = curSeq
      }
      c.set(ch, cnt - k)
      if (c.get(ch)! < k) c.delete(ch)
      dfs()
      curSeq = curSeq.slice(0, -1)
      c.set(ch, cnt)
    }
  }
};
